File: src/cpu.nim
import strutils
import rom
import ram

type REGS = string

type CPU* = object
    zero*:  int16
    t0*:    int16
    t1*:    int16
    t2*:    int16
    t3*:    int16
    t4*:    int16
    t5*:    int16
    rt*:    int16
    gp*:    int16
    sp*:    int16
    fp*:    int16
    sc*:    int16
    ra*:    int16
    pc*:    int16
    ir*:    string
    rom*:   ROM
    ram*:   RAM

proc newCPU*(program: seq[seq[string]]): CPU =
    result.ram = newRam()
    result.rom = newRom()
    var gp: int = result.rom.loadProgram(program)
    result.gp = int16(gp)
    result.pc = 0
    return result

proc setRegister(self: var CPU, dest: REGS, value: int16) = 
    case dest
    of "$t0":
        self.t0 = value 
    of "$t1":
        self.t1 = value 
    of "$t2":
        self.t2 = value 
    of "$t3":
        self.t3 = value 
    of "$t4":
        self.t4 = value 
    of "$t5":
        self.t5 = value 
    of "$pc":
        self.pc = value 
    of "$zero":
        self.zero = 0
    of "$sp":
        self.sp = value 
    of "$fp":
        self.fp = value
    of "$sc":
        self.sc = value
    of "$ra":
        self.ra = value
    of "$rt":
        self.rt = value

proc getRegister(self: var CPU, dest: REGS): int16 = 
    case dest
    of "$t0":
        return self.t0
    of "$t1":
        return self.t1 
    of "$t2":
        return self.t2 
    of "$t3":
        return self.t3 
    of "$t4":
        return self.t4 
    of "$t5":
        return self.t5 
    of "$pc":
        return self.pc 
    of "$zero":
        return 0
    of "$sp":
        return self.sp 
    of "$fp":
        return self.fp 
    of "$sc":
        return self.sc 
    of "$ra":
        return self.ra
    of "$rt":
        return self.rt

proc add(self: var CPU, dest: REGS, src1: REGS, src2: REGS) =
    let v1: int16 = self.getRegister(src1)
    let v2: int16 = self.getRegister(src2)
    self.setRegister(dest, v1 + v2)

proc sub(self: var CPU, dest: REGS, src1: REGS, src2: REGS) =
    let v1: int16 = self.getRegister(src1)
    let v2: int16 = self.getRegister(src2)
    self.setRegister(dest, v1 - v2)

proc mult(self: var CPU, dest: REGS, src1: REGS, src2: REGS) =
    let v1: int16 = self.getRegister(src1)
    let v2: int16 = self.getRegister(src2)
    self.setRegister(dest, v1 * v2)

proc kdiv(self: var CPU, dest: REGS, src1: REGS, src2: REGS) =
    let v1: int16 = self.getRegister(src1)
    let v2: int16 = self.getRegister(src2)
    # echo v1
    # echo v2
    self.setRegister(dest, v1 div v2)

proc kand(self: var CPU, dest: REGS, src1: REGS, src2: REGS) =
    let v1: int16 = self.getRegister(src1)
    let v2: int16 = self.getRegister(src2)
    self.setRegister(dest, v1 and v2)

proc kor(self: var CPU, dest: REGS, src1: REGS, src2: REGS) =
    let v1: int16 = self.getRegister(src1)
    let v2: int16 = self.getRegister(src2)
    self.setRegister(dest, v1 or v2)

proc addi(self: var CPU, dest: REGS, src1: REGS, value: int16) =
    let v1: int16 = self.getRegister(src1)
    self.setRegister(dest, v1 + value)

proc subi(self: var CPU, dest: REGS, src1: REGS, value: int16) =
    let v1: int16 = self.getRegister(src1)
    self.setRegister(dest, v1 - value)

proc multi(self: var CPU, dest: REGS, src1: REGS, value: int16) =
    let v1: int16 = self.getRegister(src1)
    self.setRegister(dest, v1 * value)

proc divi(self: var CPU, dest: REGS, src1: REGS, value: int16) =
    let v1: int16 = self.getRegister(src1)
    self.setRegister(dest, v1 div value)

proc andi(self: var CPU, dest: REGS, src1: REGS, value: int16) =
    let v1: int16 = self.getRegister(src1)
    self.setRegister(dest, v1 and value)

proc ori(self: var CPU, dest: REGS, src1: REGS, value: int16) =
    let v1: int16 = self.getRegister(src1)
    self.setRegister(dest, v1 or value)

proc sll(self: var CPU, dest: REGS, src1: REGS, value: int16) =
    let v1: int16 = self.getRegister(src1)
    self.setRegister(dest, v1 shl value)

proc srl(self: var CPU, dest: REGS, src1: REGS, value: int16) =
    let v1: int16 = self.getRegister(src1)
    self.setRegister(dest, v1 shr value)

proc move(self: var CPU, dest: REGS, src: REGS) =
    let v1: int16 = self.getRegister(src)
    self.setRegister(dest, v1)

proc li(self: var CPU, dest: REGS, value: int16) =
    self.setRegister(dest, value)

proc la(self: var CPU, dest: REGS, value: int16) =
    self.setRegister(dest, value)

proc lrw(self: var CPU, dest: REGS, offset: int16, point: int16) =
    let half_1: int16 = int16(parseInt(self.rom.get(self.gp + point + offset)))
    let half_2: int16 = int16(parseInt(self.rom.get(self.gp + point + offset + 1)))
    let value: int16 = half_1 or half_2
    self.setRegister(dest, value)

proc lrb(self: var CPU, dest: REGS, offset: int16, point: int16) =
    let y = self.rom.get(self.gp + point + offset)
    let x = int16(parseInt(y))
    let value: int16 = x
    self.setRegister(dest, value)

proc lw(self: var CPU, dest: REGS, offset: int16, point: int16) =
    let half_1: int8 = (self.ram.get(point + offset))
    let half_2: int8 = (self.ram.get(point + offset + 1))
    var value: int16 = half_1
    value = value shl 8
    value = value or half2
    self.setRegister(dest, value)

proc lb(self: var CPU, dest: REGS, offset: int16, point: int16) =
    let value: int8 = self.ram.get(point + offset)
    self.setRegister(dest, value)

proc sw(self: var CPU, src: REGS, offset: int16, point: int16) =
    let value: int16 = self.getRegister(src)
    let half_1: int8 = cast[int8](value and 0x00FF)
    self.ram.set(half_1, offset + point)

    let half_2: int8 = cast[int8](value and 0xFF00)
    self.ram.set(half_2, offset + point + 1)

proc sb(self: var CPU, src: REGS, offset: int16, point: int16) =
    let x = self.getRegister(src)
    let value: int8 = cast[int8](x and 0x00FF)
    self.ram.set(value, offset + point)

proc syscall(self: var CPU): int =
    # exit
    if self.sc == 0 :
        return -1

    # printInt t0 
    elif self.sc == 1 :
        stdout.write(self.t0)

    # printChar t0
    elif self.sc == 2 :
        stdout.write(char(self.t0))
        # echo char(self.t0)
        # echo self.t0
    
    return 1

proc jump(self: var CPU, point: int16) = 
    self.pc = point

proc jr(self: var CPU, src: REGS) = 
    self.pc = self.getRegister(src)

proc jal(self: var CPU, point: int16) = 
    self.ra = self.pc
    self.pc = point

proc ret(self: var CPU) =
    self.pc = self.ra 

proc beq(self: var CPU, src1: REGS, src2: REGS, point: int16) =
    let v1: int16 = self.getRegister(src1)
    let v2: int16 = self.getRegister(src2)
    if v1 == v2 :
        self.pc = point

proc bne(self: var CPU, src1: REGS, src2: REGS, point: int16) =
    let v1: int16 = self.getRegister(src1)
    let v2: int16 = self.getRegister(src2)
    if v1 != v2 :
        self.pc = point

proc bgt(self: var CPU, src1: REGS, src2: REGS, point: int16) =
    let v1: int16 = self.getRegister(src1)
    let v2: int16 = self.getRegister(src2)
    if v1 > v2 :
        self.pc = point

proc bge(self: var CPU, src1: REGS, src2: REGS, point: int16) =
    let v1: int16 = self.getRegister(src1)
    let v2: int16 = self.getRegister(src2)
    if v1 >= v2 :
        self.pc = point

proc blt(self: var CPU, src1: REGS, src2: REGS, point: int16) =
    let v1: int16 = self.getRegister(src1)
    let v2: int16 = self.getRegister(src2)
    if v1 < v2 :
        self.pc = point

proc ble(self: var CPU, src1: REGS, src2: REGS, point: int16) =
    let v1: int16 = self.getRegister(src1)
    let v2: int16 = self.getRegister(src2)
    if v1 <= v2 :
        self.pc = point

proc slt(self: var CPU, dest: REGS, src1: REGS, src2: REGS) =
    let v1: int16 = self.getRegister(src1)
    let v2: int16 = self.getRegister(src2)
    if (v1 < v2) :
        self.setRegister(dest, 1)
    else :
        self.setRegister(dest, 0)

proc slti(self: var CPU, dest: REGS, src1: REGS, value: int16) =
    let v1: int16 = self.getRegister(src1)
    if (v1 < value) :
        self.setRegister(dest, 1)
    else :
        self.setRegister(dest, 0)

proc kinc(self: var CPU, dest: REGS) =
    self.setRegister(dest, self.getRegister(dest) + 1)

proc kdec(self: var CPU, dest: REGS) =
    self.setRegister(dest, self.getRegister(dest) - 1)

proc getNextInstruction*(self: var CPU): int =
    if (self.pc == self.gp) :
        return 1

    let first_half: string = self.rom.get(self.pc)
    let second_half: string = self.rom.get(self.pc + 1)
    if (first_half in ["sys", "ret"]) :
        self.ir = [first_half, second_half].join("")
    else :
        self.ir = [first_half, second_half].join(" ")

    self.pc += 2
    return 0

proc execCurrentInstruction*(self: var CPU): int =
    let tokens: seq[string] = self.ir.split(" ")
    case tokens[0]
    of "add" :
        self.add(tokens[1], tokens[2], tokens[3])
    of "sub" :
        self.sub(tokens[1], tokens[2], tokens[3])
    of "mult" :
        self.mult(tokens[1], tokens[2], tokens[3])
    of "div" :
        self.kdiv(tokens[1], tokens[2], tokens[3])
    of "addi" :
        self.addi(tokens[1], tokens[2], int16(parseint(tokens[3])))
    of "subi" :
        self.subi(tokens[1], tokens[2], int16(parseint(tokens[3])))
    of "multi" :
        self.multi(tokens[1], tokens[2], int16(parseint(tokens[3])))
    of "divi" :
        self.divi(tokens[1], tokens[2], int16(parseint(tokens[3])))
    of "move" :
        self.move(tokens[1], tokens[2])
    of "li" :
        self.li(tokens[1], int16(parseint(tokens[2])))
    of "la" :
        self.la(tokens[1], int16(parseint(tokens[2])))
    of "j" :
        self.jump(int16(parseint(tokens[1])))
    of "jal" :
        self.jal(int16(parseint(tokens[1])))
    of "jr" :
        self.jr(tokens[1])
    of "return" :
        self.ret()
    of "beq" :
        self.beq(tokens[1], tokens[2], int16(parseint(tokens[3])))
    of "bne" :
        self.bne(tokens[1], tokens[2], int16(parseint(tokens[3])))
    of "bgt" :
        self.bgt(tokens[1], tokens[2], int16(parseint(tokens[3])))
    of "bge" :
        self.bge(tokens[1], tokens[2], int16(parseint(tokens[3])))
    of "blt" :
        self.blt(tokens[1], tokens[2], int16(parseint(tokens[3])))
    of "ble" :
        self.ble(tokens[1], tokens[2], int16(parseint(tokens[3])))
    of "slt" :
        self.slt(tokens[1], tokens[2], tokens[3])
    of "slti" :
        self.slti(tokens[1], tokens[2], int16(parseint(tokens[3])))
    of "inc" :
        self.kinc(tokens[1])
    of "dec" :
        self.kdec(tokens[1])
    of "sll" :
        self.sll(tokens[1], tokens[2], int16(parseint(tokens[3])))
    of "srl" :
        self.srl(tokens[1], tokens[2], int16(parseint(tokens[3])))
    of "lrb" :
        if (tokens[3][0] == '$') :
            self.lrb(tokens[1], int16(parseint(tokens[2])), self.getRegister(tokens[3]))
        else :
            self.lrb(tokens[1], int16(parseint(tokens[2])), int16(parseint(tokens[3])))
    of "lrw" :
        if (tokens[3][0] == '$') :
            self.lrw(tokens[1], int16(parseint(tokens[2])), self.getRegister(tokens[3]))
        else :
            self.lrw(tokens[1], int16(parseint(tokens[2])), int16(parseint(tokens[3])))
    of "lw" :
        if (tokens[3][0] == '$') :
            self.lw(tokens[1], int16(parseint(tokens[2])), self.getRegister(tokens[3]))
        else :
            self.lw(tokens[1], int16(parseint(tokens[2])), int16(parseint(tokens[3])))
    of "lb" :
        if (tokens[3][0] == '$') :
            self.lb(tokens[1], int16(parseint(tokens[2])), self.getRegister(tokens[3]))
        else :
            self.lb(tokens[1], int16(parseint(tokens[2])), int16(parseint(tokens[3])))
    of "sw" :
        if (tokens[3][0] == '$') :
            self.sw(tokens[1], int16(parseint(tokens[2])), self.getRegister(tokens[3]))
        else :
            self.sw(tokens[1], int16(parseint(tokens[2])), int16(parseint(tokens[3])))
    of "sb" :
        if (tokens[3][0] == '$') :
            self.sb(tokens[1], int16(parseint(tokens[2])), self.getRegister(tokens[3]))
        else :
            self.sb(tokens[1], int16(parseint(tokens[2])), int16(parseint(tokens[3])))
    of "syscall" :
        return self.syscall()
    else :
        # echo tokens
        return -2  

    return 0

File: src/PC/cpu.go
package pc

import (
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"strings"
)

type CPU struct {
	zero int16
	t0   int16
	t1   int16
	t2   int16
	t3   int16
	t4   int16
	t5   int16
	rt   int16
	sp   int16
	fp   int16
	sc   int16
	ra   uint16
	pc   uint16
	gp   uint16
	ir   string
	rom  *ROM
	ram  *RAM
}

func NewCPU(file []string) (*CPU, error) {
	var cpu CPU

	gp, err := strconv.ParseInt(file[0], 10, 16)
	if err != nil {
		return nil, err
	}
	cpu.gp = uint16(gp)

	rom, err := NewRom(file[1:])
	if err != nil {
		return nil, err
	}
	cpu.rom = rom

	cpu.ram = NewRam()
	return &cpu, nil
}

func (cpu *CPU) SetRegister(dest string, value int16) {
	switch dest {
	case "$t0":
		cpu.t0 = value
	case "$t1":
		cpu.t1 = value
	case "$t2":
		cpu.t2 = value
	case "$t3":
		cpu.t3 = value
	case "$t4":
		cpu.t4 = value
	case "$t5":
		cpu.t5 = value
	case "$pc":
		cpu.pc = uint16(value)
	case "$zero":
		cpu.zero = 0
	case "$sp":
		cpu.sp = value
	case "$fp":
		cpu.fp = value
	case "$sc":
		cpu.sc = value
	case "$ra":
		cpu.ra = uint16(value)
	case "$gp":
		cpu.gp = uint16(value)
	case "$rt":
		cpu.rt = value
	}
}

func (cpu *CPU) GetRegister(dest string) int16 {
	switch dest {
	case "$t0":
		return cpu.t0
	case "$t1":
		return cpu.t1
	case "$t2":
		return cpu.t2
	case "$t3":
		return cpu.t3
	case "$t4":
		return cpu.t4
	case "$t5":
		return cpu.t5
	case "$pc":
		return int16(cpu.pc)
	case "$zero":
		return 0
	case "$sp":
		return cpu.sp
	case "$fp":
		return cpu.fp
	case "$sc":
		return cpu.sc
	case "$ra":
		return int16(cpu.ra)
	case "$gp":
		return int16(cpu.gp)
	case "$rt":
		return cpu.rt
	}
	panic("registrador inválido")
}

func (cpu *CPU) Add(dest, src1, src2 string) {
	v1 := cpu.GetRegister(src1)
	v2 := cpu.GetRegister(src2)
	cpu.SetRegister(dest, v1+v2)
}

func (cpu *CPU) Sub(dest, src1, src2 string) {
	v1 := cpu.GetRegister(src1)
	v2 := cpu.GetRegister(src2)
	cpu.SetRegister(dest, v1-v2)
}

func (cpu *CPU) Mult(dest, src1, src2 string) {
	v1 := cpu.GetRegister(src1)
	v2 := cpu.GetRegister(src2)
	cpu.SetRegister(dest, v1*v2)
}

func (cpu *CPU) Div(dest, src1, src2 string) {
	v1 := cpu.GetRegister(src1)
	v2 := cpu.GetRegister(src2)
	cpu.SetRegister(dest, v1/v2)
}

func (cpu *CPU) Addi(dest, src string, value int16) {
	v := cpu.GetRegister(src)
	cpu.SetRegister(dest, v+value)
}

func (cpu *CPU) Subi(dest, src string, value int16) {
	v := cpu.GetRegister(src)
	cpu.SetRegister(dest, v-value)
}

func (cpu *CPU) Multi(dest, src string, value int16) {
	v := cpu.GetRegister(src)
	cpu.SetRegister(dest, v*value)
}

func (cpu *CPU) Divi(dest, src string, value int16) {
	v := cpu.GetRegister(src)
	cpu.SetRegister(dest, v/value)
}

func (cpu *CPU) Addu(dest, src1, src2 string) {
	v1 := uint16(cpu.GetRegister(src1))
	v2 := uint16(cpu.GetRegister(src2))
	cpu.SetRegister(dest, int16(v1+v2))
}

func (cpu *CPU) Addui(dest, src string, value int16) {
	v := uint16(cpu.GetRegister(src))
	result := v + uint16(value)
	cpu.SetRegister(dest, int16(result))
}

func (cpu *CPU) Subu(dest, src1, src2 string) {
	v1 := uint16(cpu.GetRegister(src1))
	v2 := uint16(cpu.GetRegister(src2))
	cpu.SetRegister(dest, int16(v1-v2))
}

func (cpu *CPU) Subui(dest, src string, value int16) {
	v := uint16(cpu.GetRegister(src))
	result := v - uint16(value)
	cpu.SetRegister(dest, int16(result))
}

func (cpu *CPU) And(dest, src1, src2 string) {
	v1 := cpu.GetRegister(src1)
	v2 := cpu.GetRegister(src2)
	cpu.SetRegister(dest, v1&v2)
}

func (cpu *CPU) Andi(dest, src string, value int16) {
	v := cpu.GetRegister(src)
	cpu.SetRegister(dest, v&value)
}

func (cpu *CPU) Or(dest, src1, src2 string) {
	v1 := cpu.GetRegister(src1)
	v2 := cpu.GetRegister(src2)
	cpu.SetRegister(dest, v1|v2)
}

func (cpu *CPU) Ori(dest, src string, value int16) {
	v := cpu.GetRegister(src)
	cpu.SetRegister(dest, v|value)
}

func (cpu *CPU) Slt(dest, src1, src2 string) {
	if cpu.GetRegister(src1) < cpu.GetRegister(src2) {
		cpu.SetRegister(dest, 1)
	} else {
		cpu.SetRegister(dest, 0)
	}
}

func (cpu *CPU) Slti(dest, src string, value int16) {
	if cpu.GetRegister(src) < value {
		cpu.SetRegister(dest, 1)
	} else {
		cpu.SetRegister(dest, 0)
	}
}

func (cpu *CPU) Sltu(dest, src1, src2 string) {
	if cpu.GetRegister(src1) < cpu.GetRegister(src2) {
		cpu.SetRegister(dest, 1)
	} else {
		cpu.SetRegister(dest, 0)
	}
}

func (cpu *CPU) Sltui(dest, src string, value uint16) {
	if uint16(cpu.GetRegister(src)) < value {
		cpu.SetRegister(dest, 1)
	} else {
		cpu.SetRegister(dest, 0)
	}
}

func (cpu *CPU) Move(dest, src string) {
	cpu.SetRegister(dest, cpu.GetRegister(src))
}

func (cpu *CPU) Li(dest string, value int16) {
	cpu.SetRegister(dest, value)
}

func (cpu *CPU) La(dest string, value uint16) {
	cpu.SetRegister(dest, int16(value))
}

func (cpu *CPU) Jump(point uint16) {
	cpu.pc = point
}

func (cpu *CPU) Jr(src string) {
	cpu.pc = uint16(cpu.GetRegister(src))
}

func (cpu *CPU) Jal(point uint16) {
	cpu.ra = cpu.pc
	cpu.pc = point
}

func (cpu *CPU) Ret() {
	cpu.pc = cpu.ra
}

func (cpu *CPU) Beq(src1, src2 string, point uint16) {
	if cpu.GetRegister(src1) == cpu.GetRegister(src2) {
		cpu.pc = point
	}
}

func (cpu *CPU) Bne(src1, src2 string, point uint16) {
	if cpu.GetRegister(src1) != cpu.GetRegister(src2) {
		cpu.pc = point
	}
}

func (cpu *CPU) Bgt(src1, src2 string, point uint16) {
	if cpu.GetRegister(src1) > cpu.GetRegister(src2) {
		cpu.pc = point
	}
}

func (cpu *CPU) Bge(src1, src2 string, point uint16) {
	if cpu.GetRegister(src1) >= cpu.GetRegister(src2) {
		cpu.pc = point
	}
}

func (cpu *CPU) Blt(src1, src2 string, point uint16) {
	if cpu.GetRegister(src1) < cpu.GetRegister(src2) {
		cpu.pc = point
	}
}

func (cpu *CPU) Ble(src1, src2 string, point uint16) {
	if cpu.GetRegister(src1) <= cpu.GetRegister(src2) {
		cpu.pc = point
	}
}

func (cpu *CPU) Inc(dest string) {
	cpu.SetRegister(dest, cpu.GetRegister(dest)+1)
}

func (cpu *CPU) Dec(dest string) {
	cpu.SetRegister(dest, cpu.GetRegister(dest)-1)
}

func (cpu *CPU) Rand(dest string) {
	cpu.SetRegister(dest, int16(rand.Intn(65536)-32768))
}

func (cpu *CPU) Lrb(dest string, offset, point uint16) {
	addr := cpu.gp + point + offset
	v, err := strconv.Atoi(cpu.rom.rom[addr])
	if err != nil {
		panic(err)
	}

	cpu.SetRegister(dest, int16(v))
}

func (cpu *CPU) Lrw(dest string, offset, point uint16) {
	addr := cpu.gp + point + offset

	h1, err := strconv.Atoi(cpu.rom.rom[addr])
	if err != nil {
		panic(err)
	}

	h2, err := strconv.Atoi(cpu.rom.rom[addr+1])
	if err != nil {
		panic(err)
	}

	value := (int16(h1) << 8) | int16(h2)
	// print(value)
	cpu.SetRegister(dest, value)
}

func (cpu *CPU) Sll(dest, src string, value uint16) {
	v := int16(cpu.GetRegister(src))
	cpu.SetRegister(dest, int16(v<<value))
}

func (cpu *CPU) Srl(dest, src string, value uint16) {
	v := int16(cpu.GetRegister(src))
	cpu.SetRegister(dest, int16(v>>value))
}

func (cpu *CPU) Sb(src string, offset int16, reg_index string) {
	value := uint8(cpu.GetRegister(src))
	index := int32(cpu.GetRegister(reg_index)) + int32(offset)
	cpu.ram.save_byte(value, uint16(index))
}

func (cpu *CPU) Sw(src string, offset int16, reg_index string) {
	value := uint16(cpu.GetRegister(src))
	index := int32(cpu.GetRegister(reg_index)) + int32(offset)
	cpu.ram.save_world(value, uint16(index))
}

func (cpu *CPU) Lb(dest string, offset int16, reg_index string) {
	index := int32(cpu.GetRegister(reg_index)) + int32(offset)
	value := cpu.ram.load_byte(uint16(index))
	cpu.SetRegister(dest, int16(value))
}

func (cpu *CPU) Lw(dest string, offset int16, reg_index string) {
	index := int32(cpu.GetRegister(reg_index)) + int32(offset)
	value := cpu.ram.load_world(uint16(index))
	cpu.SetRegister(dest, int16(value))
}

// 0 - exit
// 1 - print t0 como int8
// 2 - print t0 como utin8
// 3 - print t0 como int16
// 4 - print t0 como uint16
// 5 - print t0 como char utf8
func (cpu *CPU) Syscall() {
	value := cpu.t0
	switch cpu.sc {
	case 0:
		os.Exit(0)
	case 1:
		print(int8(value))
	case 2:
		print(uint8(value))
	case 3:
		print(value)
	case 4:
		print(uint16(value))
	case 5:
		fmt.Printf("%c", int8(value))

	// SYSCALLS para testar o computador
	case 1001:
		println(int8(value))
	case 1002:
		println(uint8(value))
	case 1003:
		println(value)
	case 1004:
		println(uint16(value))
	case 1005:
		fmt.Printf("%c\n", int8(value))

	}
}

func (cpu *CPU) ExecCurrentInstruction(tokens []string) int {
	switch tokens[0] {

	case "add":
		cpu.Add(tokens[1], tokens[2], tokens[3])

	case "sub":
		cpu.Sub(tokens[1], tokens[2], tokens[3])

	case "mult":
		cpu.Mult(tokens[1], tokens[2], tokens[3])

	case "div":
		cpu.Div(tokens[1], tokens[2], tokens[3])

	case "addi":
		v, _ := strconv.Atoi(tokens[3])
		cpu.Addi(tokens[1], tokens[2], int16(v))

	case "subi":
		v, _ := strconv.Atoi(tokens[3])
		cpu.Subi(tokens[1], tokens[2], int16(v))

	case "multi":
		v, _ := strconv.Atoi(tokens[3])
		cpu.Multi(tokens[1], tokens[2], int16(v))

	case "divi":
		v, _ := strconv.Atoi(tokens[3])
		cpu.Divi(tokens[1], tokens[2], int16(v))

	case "move":
		cpu.Move(tokens[1], tokens[2])

	case "li":
		v, _ := strconv.Atoi(tokens[2])
		cpu.Li(tokens[1], int16(v))

	case "la":
		v, _ := strconv.Atoi(tokens[2])
		cpu.La(tokens[1], uint16(v))

	case "j":
		v, _ := strconv.Atoi(tokens[1])
		cpu.Jump(uint16(v))

	case "jal":
		v, _ := strconv.Atoi(tokens[1])
		cpu.Jal(uint16(v))

	case "jr":
		cpu.Jr(tokens[1])

	case "return":
		cpu.Ret()

	case "beq":
		v, _ := strconv.Atoi(tokens[3])
		cpu.Beq(tokens[1], tokens[2], uint16(v))

	case "bne":
		v, _ := strconv.Atoi(tokens[3])
		cpu.Bne(tokens[1], tokens[2], uint16(v))

	case "bgt":
		v, _ := strconv.Atoi(tokens[3])
		cpu.Bgt(tokens[1], tokens[2], uint16(v))

	case "bge":
		v, _ := strconv.Atoi(tokens[3])
		cpu.Bge(tokens[1], tokens[2], uint16(v))

	case "blt":
		v, _ := strconv.Atoi(tokens[3])
		cpu.Blt(tokens[1], tokens[2], uint16(v))

	case "ble":
		v, _ := strconv.Atoi(tokens[3])
		cpu.Ble(tokens[1], tokens[2], uint16(v))

	case "slt":
		cpu.Slt(tokens[1], tokens[2], tokens[3])

	case "slti":
		v, _ := strconv.Atoi(tokens[3])
		cpu.Slti(tokens[1], tokens[2], int16(v))

	case "sltu":
		cpu.Sltu(tokens[1], tokens[2], tokens[3])

	case "sltui":
		v, _ := strconv.Atoi(tokens[3])
		cpu.Sltui(tokens[1], tokens[2], uint16(v))

	case "and":
		cpu.And(tokens[1], tokens[2], tokens[3])

	case "andi":
		v, _ := strconv.Atoi(tokens[3])
		cpu.Andi(tokens[1], tokens[2], int16(v))

	case "or":
		cpu.Or(tokens[1], tokens[2], tokens[3])

	case "ori":
		v, _ := strconv.Atoi(tokens[3])
		cpu.Ori(tokens[1], tokens[2], int16(v))

	case "sll":
		v, _ := strconv.Atoi(tokens[3])
		cpu.Sll(tokens[1], tokens[2], uint16(v))

	case "srl":
		v, _ := strconv.Atoi(tokens[3])
		cpu.Srl(tokens[1], tokens[2], uint16(v))

	case "inc":
		cpu.Inc(tokens[1])

	case "dec":
		cpu.Dec(tokens[1])

	case "rand":
		cpu.Rand(tokens[1])

	case "lrb":
		var offset uint16
		if tokens[2][0] == '$' {
			offset = uint16(cpu.GetRegister(tokens[2]))
		} else {
			v, _ := strconv.Atoi(tokens[2])
			offset = uint16(v)
		}

		if tokens[3][0] == '$' {
			cpu.Lrb(tokens[1], uint16(offset), uint16(cpu.GetRegister(tokens[3])))
		} else {
			point, _ := strconv.Atoi(tokens[3])
			cpu.Lrb(tokens[1], uint16(offset), uint16(point))
		}

	case "lrw":
		offset, _ := strconv.Atoi(tokens[2])

		if tokens[3][0] == '$' {
			cpu.Lrw(tokens[1], uint16(offset), uint16(cpu.GetRegister(tokens[3])))
		} else {
			point, _ := strconv.Atoi(tokens[3])
			cpu.Lrw(tokens[1], uint16(offset), uint16(point))
		}

	case "sb":
		offset, _ := strconv.Atoi(tokens[2])
		cpu.Sb(tokens[1], int16(offset), tokens[3])

	case "sw":
		offset, _ := strconv.Atoi(tokens[2])
		cpu.Sw(tokens[1], int16(offset), tokens[3])

	case "lb":
		offset, _ := strconv.Atoi(tokens[2])
		cpu.Lw(tokens[1], int16(offset), tokens[3])

	case "lw":
		offset, _ := strconv.Atoi(tokens[2])
		cpu.Lw(tokens[1], int16(offset), tokens[3])

	case "syscall":
		cpu.Syscall()

	default:
		fmt.Printf("Instrução invalida \"%s\".\n", tokens[0])
		return -2
	}

	return 0
}

func (cpu *CPU) CarryNextInstruction() {
	h1 := cpu.rom.rom[cpu.pc]
	h2 := cpu.rom.rom[cpu.pc+1]
	if h1 == "sys" || h1 == "ret" {
		cpu.ir = h1 + h2
	} else {
		cpu.ir = h1 + " " + h2
	}
	cpu.pc += 2
}

func (cpu *CPU) Run() {
	for {
		cpu.CarryNextInstruction()
		ok := cpu.ExecCurrentInstruction(strings.Split(cpu.ir, " "))
		if ok != 0 {
			return
		}

		if cpu.pc >= cpu.gp {
			return
		}
		// time.Sleep(10000000)
	}
}

func (cpu *CPU) ShowRom() {
	for i, v := range cpu.rom.rom {
		fmt.Printf("%d -> %s\n", i, v)
		if i == 150 {
			return
		}
	}
}
